### ISSUE TITLE
Batching when using multiple datasets

### ISSUE THREAD

**Issue by humzaiqbal** [mlfoundations/open_clip#762]
If I'm running using multiple web datasets can I expect that a single batch which I feed to the model would contain examples from just one dataset or multiple datasets. Been poking around and it seems like the answer is just a single dataset as it seems like a batch would primarily consist of shards (in my case my batch size is much smaller than the number of text image pairs in a shard so just a single shard). Does that sound right?

**Comment by gabrielilharco** [comment]
Internally the dataloader uses a shuffle buffer with fixed size. Once a worker has a shard, it pulls samples from the shard into the shuffle buffer, and samples are randomly drawn from the shuffle buffer. If you have very big shards (or small buffer size), then most of the times each worker will produce batches from the same dataset. However, if you have multiple workers, then there's also some mixing on the global batch

**Comment by rwightman** [comment]
there's also a shuffle buffer that, depending on the size relative the per-shard # samples can mix samples across neighbouring shards...

**Comment by abhishek0093** [comment]
Hi @rwightman , @gabrielilharco 
Thanks a lot for this amazing work and implementation! I'm having a similar problem. I’m trying to evaluate the trained model on two separate dataset. The way I was thinking is two concatenate the two dataset like this `--val-data "/root/img_dataset_1/{00000..00005}.tar::/root/img_dataset_2/{00000..00005}.tar"` with each tar having 1k examples. 
Earlier I was expecting to get the embeddings in the same order so that I could handle metrices in get_clip_metrics function something like this : 
```
def get_clip_metrics(image_features, text_features, logit_scale):

    cutoff_examples = 6000
    metrics = {}

   # Handle the 1st dataset 
    logits_per_image = (logit_scale * image_features[:cutoff_examples] @ text_features[:cutoff_examples].t()).detach().cpu()
    logits_per_text = logits_per_image.t().detach().cpu()
    logits = {"case1_img2txt": logits_per_image, "case1_txt2img": logits_per_text}
    ground_truth = torch.arange(len(text_features[:cutoff_examples])).view(-1, 1)

    for name, logit in logits.items():
        ranking = torch.argsort(logit, descending=True)
        preds = torch.where(ranking == ground_truth)[1]
        preds = preds.detach().cpu().numpy()
        metrics[f"{name}_mean_rank"] = preds.mean() + 1
        metrics[f"{name}_median_rank"] = np.floor(np.median(preds)) + 1
        for k in [1, 5, 10]:
            metrics[f"{name}_R@{k}"] = np.mean(preds < k)

    # Handle the 2nd dataset
    logits_per_image = (logit_scale * image_features[cutoff_examples:] @ text_features[cutoff_examples:].t()).detach().cpu()
    logits_per_text = logits_per_image.t().detach().cpu()
    logits = {"case2_img2txt": logits_per_image, "case2_txt2img": logits_per_text}
    ground_truth = torch.arange(len(text_features[cutoff_examples:])).view(-1, 1)

    for name, logit in logits.items():
        ranking = torch.argsort(logit, descending=True)
        preds = torch.where(ranking == ground_truth)[1]
        preds = preds.detach().cpu().numpy()
        metrics[f"{name}_mean_rank"] = preds.mean() + 1
        metrics[f"{name}_median_rank"] = np.floor(np.median(preds)) + 1
        for k in [1, 5, 10]:
            metrics[f"{name}_R@{k}"] = np.mean(preds < k)

    return metrics

```
But as it turned out the embeddings aren’t sequential when using num_workers > 1. I’m guessing it’s because of this line in data.py file. 
```
 pipeline.extend([
            wds.split_by_worker,
            # at this point, we have an iterator over the shards assigned to each worker
            wds.tarfile_to_samples(handler=log_and_continue),
        ])
```
when we define split across workers for not train(validation) case, but I’m not sure. Is there some way where I can do separate validation on two datasets and get metrics as above ? Very thanks for the help.

**Comment by abhishek0093** [comment]
Nvm, I think my approach above was wrong to append the new dataset to the 1st dataset and expect in order. I created a new argument called --val_data2 and handled it similar to the --val_data argument for the 2nd dataset validation and now it is working good. Thanks everyone for the help.
